c/tskit/trees.c

```diff
@@ -2103,22 +2126,635 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+    }
+out:
+    tsk_safe_free(mut_samples);
+    tsk_safe_free(node_paths);
+    tsk_safe_free(stack);
+
+    return ret;
+}
+
+static int
+norm_hap_weighted(tsk_size_t state_dim, const double *hap_weights,
+    tsk_size_t TSK_UNUSED(n_a), tsk_size_t TSK_UNUSED(n_b), double *result, void *params)
+{
+    sample_count_stat_params_t args = *(sample_count_stat_params_t *) params;
+    const double *weight_row;
+    double n;
+    for (tsk_size_t k = 0; k < state_dim; k++) {
+        weight_row = GET_2D_ROW(hap_weights, 3, k);
+        n = (double) args.sample_set_sizes[k];
+        // TODO: what to do when n = 0
+        result[k] = weight_row[0] / n;
+    }
+    return 0;
+}
+
+static int
+norm_total_weighted(tsk_size_t state_dim, const double *TSK_UNUSED(hap_weights),
+    tsk_size_t n_a, tsk_size_t n_b, double *result, void *TSK_UNUSED(params))
+{
+    for (tsk_size_t k = 0; k < state_dim; k++) {
+        result[k] = 1 / (double) (n_a * n_b);
+    }
+    return 0;
+}
+
+static void
+get_all_samples_bits(tsk_bit_array_t *all_samples, tsk_size_t n, tsk_size_t n_chunks)
+{
+    const tsk_bit_array_t all = ~((tsk_bit_array_t) 0);
+    const tsk_bit_array_t remainder_samples = n % TSK_BIT_ARRAY_NUM_BITS;
+    all_samples[n_chunks - 1] = remainder_samples ? ~(all << remainder_samples) : all;
+    for (tsk_size_t i = 0; i < n_chunks - 1; i++) {
+        all_samples[i] = all;
+    }
+}
+
+static void
+sample_weights_to_bit_array(const double *weights, tsk_size_t num_sample_sets,
+    tsk_size_t num_samples, tsk_size_t num_sample_chunks, tsk_bit_array_t **sample_bits)
+{
+    const double *weight_row;
+    tsk_bit_array_t *bits_row;
+    for (tsk_bit_array_t i = 0; i < num_samples; i++) {
+        weight_row = GET_2D_ROW(weights, num_sample_sets, i);
+        for (tsk_size_t k = 0; k < num_sample_sets; k++) {
             if (weight_row[k] != 0) {
-                ret = TSK_ERR_DUPLICATE_SAMPLE;
+                bits_row = GET_2D_ROW(*sample_bits, num_sample_chunks, k);
+                tsk_add_bit_to_bit_array(bits_row, i);
+            }
+        }
+    }
+}
+
+typedef int norm_func_t(tsk_size_t state_dim, const double *hap_weights, tsk_size_t n_a,
+    tsk_size_t n_b, double *result, void *params);
+
+// TODO: should site_a/b be an id_t?
+static int
+compute_general_two_site_stat_result(tsk_size_t site_a, tsk_size_t site_a_offset,
+    tsk_size_t site_b, tsk_size_t site_b_offset, tsk_size_t num_sample_chunks,
+    const tsk_size_t *num_alleles, const tsk_bit_array_t *state, tsk_size_t state_dim,
+    tsk_bit_array_t *sample_sets, tsk_size_t result_dim, general_stat_func_t *f,
+    sample_count_stat_params_t *f_params, norm_func_t *norm_f, bool polarised,
+    double *result)
+{
+    int ret = 0;
+    const tsk_bit_array_t *A_samples, *B_samples;
+    tsk_size_t w_A = 0, w_B = 0, w_AB = 0;
+    tsk_bit_array_t *ss_row; // ss_ prefix refers to a sample set
+    tsk_bit_array_t ss_A_samples[num_sample_chunks], ss_B_samples[num_sample_chunks],
```

Comment:
I don't think we can allocate on the stack like this, we'll eventually overflow for large sample sizes, and it's very hard to tell in advance when it's going to happen. It's just not worth the trouble.

Reply:
(You might notice that we do a lot of allocing from the stack in the test suite - but we can get away with a lot of stuff there if it makes the code easier to write and read because the inputs are fixed)